### PR TITLE
Simplify use of locks in our Vulkan layer's SubmissionTracker class

### PR DESCRIPTION
**Pull request by ronaldfw**
We simplify the use of locks in the SubmissionTracker class to make it
easier to reason about correctness of the code and avoid sometimes
leaking query slots.

For example, in methods like MarkCommandBufferBegin, the mutex was
locked, then freed again, then it was checked if we were capturing,
then the mutex was locked again to record the timestamp. It was possible
that capturing was stopped in the middle and OnCaptureFinished called.
This lead to query slots that should have been cleaned up in
OnCaptureFinished to not get cleaned up at all.

We also tighten the check on whether we are capturing and directly tie
this check to the calls to OnCaptureFinished. The previous use of the
IsCapturing() method sometimes led to imcomplete command buffer states
that were neither cleaned up in OnCaptureFinished nor properly
completed.

With this change, all query slots are now properly cleaned up after the 
call to OnCaptureFinished and after all remaining submits have been 
completed.

Tested: Manually ran Vulkan layer with two sample applications.
Bug: http://b/181097789

### CODE REVIEW

File: src/OrbitVulkanLayer/SubmissionTracker.h [hunk 1]
```diff
@@ -218,48 +219,44 @@ class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
...
-      absl::WriterMutexLock lock(&mutex_);
       CHECK(command_buffer_to_state_.contains(command_buffer));
       command_buffer_to_state_.at(command_buffer).command_buffer_begin_slot_index =
           std::make_optional(slot_index);
     }
   }
 
   void MarkCommandBufferEnd(VkCommandBuffer command_buffer) {
-    if (!IsCapturing()) {
+    absl::WriterMutexLock lock(&mutex_);
```
Comment (florian-kuebler) [inline src/OrbitVulkanLayer/SubmissionTracker.h:238]: Just to mention that with the argument below, a read lock would be fine, but agree to make it simpler to reason about.

File: src/OrbitVulkanLayer/SubmissionTracker.h [hunk 2]
```diff
@@ -499,6 +494,7 @@ class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
   // This method also resets all the timer slots that have been read.
   // It is assumed to be called periodically, e.g. on `vkQueuePresentKHR`.
   void CompleteSubmits(VkDevice device) {
+    absl::WriterMutexLock lock(&mutex_);
```
Comment (florian-kuebler) [inline src/OrbitVulkanLayer/SubmissionTracker.h:497]: Should we really lock the complete present? I guess it is fine for now, however.

**Review by florian-kuebler — APPROVED** [review]
LGTM.

File: src/OrbitVulkanLayer/SubmissionTracker.h (hunk 1 above)
Comment (ronaldfw) [reply src/OrbitVulkanLayer/SubmissionTracker.h:238]: This is writing command_buffer_end_slot_index, which may also be written / reset by OnCaptureFinished at the same time. Or am I missing something?

Comment (florian-kuebler) [reply src/OrbitVulkanLayer/SubmissionTracker.h:238]: Thanks. I was missing this.

File: src/OrbitVulkanLayer/SubmissionTracker.h (hunk 2 above)
Comment (ronaldfw) [reply src/OrbitVulkanLayer/SubmissionTracker.h:497]: Yes, I think the scope here can be a bit tighter, but the write lock is at least needed for the PullCompletedSubmissions call (it was there before) and the first loop here (which is potentially adding submissions back into the queue). I think I'll keep it as-is for simplicity but we can definitely keep this in mind for an improvement.

Comment (florian-kuebler) [reply src/OrbitVulkanLayer/SubmissionTracker.h:497]: Sounds good.

**Pull request merged**